### PR TITLE
MAT-6569: Human Readable use Stratification Description

### PR DESCRIPTION
**Pull request by ethankaplan**
## CQL to ELM Translation Service PR

Jira Ticket: [MAT-6569](https://jira.cms.gov/browse/MAT-6569)
(Optional) Related Tickets:

### All Submissions
* [ ] This PR has the JIRA linked.
* [ ] Required tests are included.
* [ ] No extemporaneous files are included (i.e Complied files or testing results).
* [ ] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [ ] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java [hunk 1]
```diff
@@ -142,7 +142,8 @@ HumanReadableMeasureInformationModel buildMeasureInfo(Measure measure) {
                 ? measure.getMeasureMetaData().getSteward().getName()
                 : null)
         .measureTypes(HumanReadableUtil.getMeasureTypes(measure))
-        .stratification(HumanReadableUtil.getStratification(measure))
+        // .stratification(HumanReadableUtil.getStratification(measure))
```
Comment (RohitKandimalla) [inline src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: We don't need this field ?

Comment (jkotanchik-SB) [inline src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: Dead code can be removed.

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java [hunk 2]
```diff
@@ -71,6 +71,17 @@ public static String getStratification(Measure measure) {
     return null;
   }
 
+  public static String getStratificationDescription(Measure measure) {
+    if (CollectionUtils.isNotEmpty(measure.getGroups())) {
+      for (Group group : measure.getGroups()) {
+        if (CollectionUtils.isNotEmpty(group.getStratifications())) {
+          return group.getStratifications().get(0).getDescription();
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java]: Do we need to worry about multiple stratifications in a group?

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java [hunk 3]
```diff
@@ -71,6 +71,17 @@ public static String getStratification(Measure measure) {
...
+        }
```
Comment (RohitKandimalla) [inline src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java:75]: Why are we returning only the first description, Shouldn't this be a concatenation of all stratification descriptions ?

Comment (RohitKandimalla) [reply src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java:75]: Also Nice to have few unit test cases

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java (hunk 1 above)
Comment (RohitKandimalla) [reply src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: Looks like HumanReadableMeasureInformationModel.stratification is the cqlDefinitions assigned, and HumanReadableMeasureInformationModel.stratificationDescription holds the information about description.

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java (hunk 3 above)
Comment (ethankaplan) [reply src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java:75]: There's only going to be one description, covering all of the stratifications

Comment (RohitKandimalla) [reply src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java:75]: Given: A QDM Measure
When: on the Population Criteria tab, on a Criteria sub-tab, in the Stratifications section
Then: There is a Stratification Description field for every Stratification field (same layout as for QI-Core measures)

May be we need to get some clarification, The 1st AC reads different.

Comment (ethankaplan) [reply src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java:75]: My previous comment is wrong, AC was changed to reflect multiple.

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java [hunk 4]
```diff
@@ -58,14 +58,24 @@ public static List<String> getMeasureTypes(Measure measure) {
   }
 
   public static String getStratification(Measure measure) {
+    // Collects and returns all stratification descriptions for display
     if (CollectionUtils.isNotEmpty(measure.getGroups())) {
+      String allDescriptions = "";
       for (Group group : measure.getGroups()) {
+        if (allDescriptions.length() > 0
+            && CollectionUtils.isNotEmpty(group.getStratifications())) {
+          allDescriptions += "\n";
```
Comment (RohitKandimalla) [inline src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java]: It will be good to have few unit test cases to verify multiple groups and stratifications.

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java [hunk 5]
```diff
@@ -58,14 +58,24 @@ public static List<String> getMeasureTypes(Measure measure) {
   }
 
   public static String getStratification(Measure measure) {
+    // Collects and returns all stratification descriptions for display
     if (CollectionUtils.isNotEmpty(measure.getGroups())) {
+      String allDescriptions = "";
       for (Group group : measure.getGroups()) {
+        if (allDescriptions.length() > 0
```
Comment (RohitKandimalla) [inline src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java]:
```suggestion
        if (!allDescriptions.isEmpty()
```
Since allDescriptions is a string, we could use isEmpty. It does the same thing but helps developers to read better.

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java [hunk 6]
```diff
@@ -58,14 +58,24 @@ public static List<String> getMeasureTypes(Measure measure) {
   }
 
   public static String getStratification(Measure measure) {
+    // Collects and returns all stratification descriptions for display
     if (CollectionUtils.isNotEmpty(measure.getGroups())) {
+      String allDescriptions = "";
```
Comment (RohitKandimalla) [inline src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java]: Strings in JAVA are immutable, so every time we modify the original string it creates a new copy, which is unnecessary in our case. May be we could use StringBuilder ? StringBuilder also helps us with safely append a sub-string.

**Review by RohitKandimalla — DISMISSED** [review]

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java (hunk 6 above)
Comment (jkotanchik-SB) [reply src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java]: Agree.  Though, concatenating a few short Strings is perfectly fine (e.g., a log output or error message).

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java (hunk 4 above)
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java]: This feels unnecessary since `Collectors.joining()` below is doing the same thing, but single space. If we want double space, then the joining argument could be changed.

Comment (ethankaplan) [reply src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java]: This adds a new line when it goes to a new, nonempty group. Otherwise it will go (2 in first group, one in second):
```
Strat1
Strat2Strat3
```

Comment (ethankaplan) [reply src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java]: But this might not matter anyway since the method of building the output is changing

Comment (jkotanchik-SB) [reply src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java]: The newline can be moved to right after the `allDescriptions` append and then `trim()` the extra newline on the return.

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java [hunk 7]
```diff
@@ -58,14 +58,23 @@ public static List<String> getMeasureTypes(Measure measure) {
...
+              .append(
+                  group.getStratifications().stream()
+                      .map(strat -> strat.getDescription())
+                      .collect(Collectors.joining("\n")))
+              .append("\n");
         }
       }
+      if (!allDescriptions.isEmpty()) {
+        return allDescriptions.toString().trim();
+      }
```
Comment (adongare) [inline src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java]: nice to escape for html entities for the safe side. you can use  HumanReadableUtil.escapeHtmlString
```suggestion
     return HumanReadableUtil.escapeHtmlString(allDescriptions.toString());
```

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java [hunk 8]
```diff
@@ -58,14 +58,23 @@ public static List<String> getMeasureTypes(Measure measure) {
...
         if (CollectionUtils.isNotEmpty(group.getStratifications())) {
-          return group.getStratifications().stream()
-              .map(strat -> strat.getCqlDefinition())
-              .collect(Collectors.joining("\n"));
+          allDescriptions
+              .append(
+                  group.getStratifications().stream()
+                      .map(strat -> strat.getDescription())
+                      .collect(Collectors.joining("\n")))
+              .append("\n");
```
Comment (adongare) [inline src/main/java/gov/cms/mat/cql_elm_translation/utils/HumanReadableUtil.java]:
```suggestion
          allDescriptions
              .append(
                  group.getStratifications().stream()
                      .map(Stratification::getDescription)
                      .filter(StringUtils::isNotBlank)
                      .collect(Collectors.joining("\n")))
              .append("\n");
```

**Review by adongare — APPROVED** [review]
